Hardware ExUnit Tests

mobileoverlord opened the pull request:
Compiling the system on travis provides a pretty warm and fuzzy feeling just knowing that it succeeds. This PR takes that one step further by producing test firmware and uploading the result to an S3 bucket. Here is an overview of how this works.

The system is built and `mix firmware` is called in `MIX_ENV=test` which brings in the `nerves_system_test` dependency and produces an application. This application knows how to do the following:

* Execute shared system ExUnit tests located in the priv dir of `nerves_system_test`
* Capture the IO and result map
* Connect to the device channel of a running instance of the `nerves_test_server` [server](https://github.com/mobileoverlord/nerves_test_server)
* Report the test results
* Wait for new firmware

When a device establishes a connection with the server it identifies the system it is running. In this case its `nerves_system_rpi0`. It will only be asked to run firmware for that system. In the future, it is my plan to allow the test server to add attribute tags to devices and associate build matrices to repositories so we can test the same firmware on multiple _types_ of boards which all use the same system. For example, `nerves_system_rpi0` should be tested on Raspberry Pi Zero and Zero W.

With the way it stands, the changes here should not affect any users running in dev / prod. We may need to add `optional: false` to the `nerves_test_server` dependency so it is not included automatically when the system is built in env test in a parent project.

Let me know what you all think.

fhunleth commented on config/config.exs:
```diff
@@ -0,0 +1,67 @@
+  ]
+
+config :nerves_runtime, :kv,
+  nerves_fw_application_part0_devpath: "/dev/mmcblk0p3",
```
It would be great if these variables could be deleted. It looks like they're copy/paste from the defaults in the `fwup.conf` file and it would be better to not have to duplicate them. I don't think that I'd even want some of them duplicated, since it defeats the purpose of code introspecting the firmware to figure these things out.

mobileoverlord replied:
Yeah, I didn't like the fact that I had to copy them there, but, they are only used to exercise the test located here: https://github.com/mobileoverlord/nerves_system_test/blob/master/priv/test/nerves_runtime_test.exs#L15

I needed an outside way to validate that the introspection is working.

fhunleth commented on mix.exs:
```diff
@@ -13,17 +20,28 @@ defmodule NervesSystemRpi0.Mixfile do
      package: package(),
      deps: deps(),
-     aliases: ["deps.precompile": ["nerves.env", "deps.precompile"]]]
+     aliases: ["deps.precompile": ["nerves.env", "nerves.precompile", "deps.precompile"],
```
What's `nerves.precompile`?

fhunleth replied:
Oops. This is the normal `mix` hack to make things work. I think that I missed when this line changed into two additional items for everything.

fhunleth commented:
Does this get rid of the trivial `.fw` file that boots to an Erlang prompt? That one is pretty important when debugging Linux and Erlang VM issues since it cuts out a lot of code from obfuscating issues.

fhunleth commented:
Just to add to the comment, I'm feeling like there's at least two `.fw` files that need to get built. One is the trivial one that boots to an Erlang prompt, the second is the one that you're working on that integrates with the official Nerves test report server. I could imagine us adding a demo app or test app that runs through a battery of tests but isn't coupled to the official report server. That latter one would be good for users so that they could test their changes without creating PRs to get their changes run on our test farm.

electricshaman commented:
Is this just for system tests?

mobileoverlord commented:
> Ok, this is a super cool PR

@fhunleth Thanks for taking a peek. 

> could you break this up into multiple commits? Some of the commits (like the bump to OTP 20 and the rootfs_overlay rename) seem trivial to merge and we'd want them right away

Absolutely. While putting them together I noticed that they were leaking in. I'll pull them out and PR them separately.

> Does this get rid of the trivial .fw file that boots to an Erlang prompt? That one is pretty important when debugging Linux and Erlang VM issues since it cuts out a lot of code from obfuscating issues.

It does not. That fw file is still being created and uploaded to S3. I agree that those fw files are important and we should work on making them more visible. Possibly through the UI that we are building for the Nerves test server. Those files are being uploaded with [this S3 task](https://github.com/nerves-project/nerves_system_rpi0/blob/system_test/.travis.yml#L73-L83). Those files are already publicly available, for example:

https://s3.amazonaws.com/nerves/test/nerves_system_rpi0-master.fw

These files are scheduled to expire in 1 week from the time it was created.

> I'm feeling like there's at least two .fw files that need to get built. One is the trivial one that boots to an Erlang prompt, the second is the one that you're working on that integrates with the official Nerves test report server.

Yup, totally agree

> I could imagine us adding a demo app or test app that runs through a battery of tests but isn't coupled to the official report server. That latter one would be good for users so that they could test their changes without creating PRs to get their changes run on our test farm.

I believe that what you are saying is that we should take `nerves_system_test` and split out the component which reports the results back to the report server. We could then include `nerves_system_test` by default when the system is being built in `Mix.env=:test` . I like this proposal, it makes more sense overall.

Its interesting to mention that the all of this is open source and the idea is that anyone out there can implement their own instance of the test server publicly or privately. This way, individuals and companies can run their own test farms.

@electricshaman Yes, the overall intention of this is focused on executing ExUnit tests on a device and report the results back to a report server.

electricshaman commented:
@mobileoverlord Right, that's cool but I guess I'm asking what constitutes the System Under Test here?  What are we trying to pass/fail?  This makes the assumption that the system is able to boot up and execute the VM, no?

electricshaman commented:
Different line of thought, so making a separate comment.

Something similar that I think would be super valuable is having a convention for executing a particular set of tests just before the application is started, dumping the results to the tty, and then taking a configurable action based on the result.

The default pass action would be to start the application, and the default fail action could be to halt and display an error message or dump the stack to a certain location, or maybe toggle the active partition and restart the system.

In a kitting, QC, or production testing setup, one might configure the pass action to just shutdown the system once all the tests pass, whereas the fail action would keep the system up and display a error message on the tty so whoever is running the tests could set the unit aside for later triage.  

I'm mentioning this here because it might be something that could fit into the nerves_test_server umbrella and share some of that code, but it might also be its own thing.

mobileoverlord commented:
@fhunleth 
I updated the structure to include an example project at `/test` which knows how to use the system produced and create firmware that can run tests on the devices as part of the test farm. I also split some of the general tasks like fixing elixir 1.5 warnings and updating travis to use otp 20 into a separate PR. Let me know if this is what you were thinking.

mobileoverlord commented:
I've updated the project to include some changes:
* Firmware files are now signed
* Network SSID and PSK are referenced from the app data partition
* The test project is in a subproject under `test/`

The latest build ran and executed on my locally connected 0W
```elixir
%NervesTestServer.Build{__meta__: #Ecto.Schema.Metadata<:loaded, "builds">,
  device: "nerves-6086", end_time: #Ecto.DateTime<2017-08-14 16:06:18>,
  id: "78ed0f75-a8f7-4dbc-badf-13e2279d6310",
  inserted_at: ~N[2017-08-14 16:04:50.184609], org: "nerves-project",
  result: %{"failures" => 0, "skipped" => 0, "total" => 3},
  result_io: "...\n\nFinished in 2.7 seconds\n3 tests, 0 failures\n\nRandomized with seed 677275\n",
  start_time: #Ecto.DateTime<2017-08-14 16:05:21>, system: "nerves_system_rpi0",
  updated_at: ~N[2017-08-14 16:06:18.485156],
  vcs_id: "c39343caa6cf665110b1b16e060837c78f80b50f"}
```

I think that this is ready for another review

fhunleth commented on .travis.yml:
```diff
@@ -14,22 +14,30 @@ addons:
-  - make clean
-  - make
-  - cd ..
+  - wget https://github.com/fhunleth/fwup/releases/download/v${FWUP_VERSION}/fwup_${FWUP_VERSION}_amd64.deb
```
Nice!

fhunleth commented on test/README.md:
```diff
@@ -0,0 +1,29 @@
+# Test
+
```
Probably should add minimal documentation here for how to use this.

fhunleth commented on test/README.md:
```diff
@@ -0,0 +1,29 @@
+`MIX_TARGET` environment variable. If `MIX_TARGET` is unset, `mix` builds an
+image that runs on the host (e.g., your laptop). This is useful for executing
+logic tests, running utilities, and debugging. Other targets are represented by
+a short name like `rpi3` that maps to a Nerves system image for that platform.
```
rpi3 -> rpi0 ;-)

fhunleth commented on test/config/config.exs:
```diff
@@ -0,0 +1,57 @@
+    [:config, :network_interface, "wlan0", :psk]
+  ]
+
+config :nerves_network, :default,
```
Could you add docs here for why `usb0` and `eth0` are listed and what this does? I wouldn't expect them to be used for the test.

fhunleth commented on test/config/config.exs:
```diff
@@ -0,0 +1,57 @@
+  ]
+
+config :nerves_runtime, :kv,
+  nerves_fw_application_part0_devpath: "/dev/mmcblk0p3",
```
Could you add a comment here about why we need to duplicate all of this information?

fhunleth commented on test/priv/test/test_test.exs:
```diff
@@ -0,0 +1,8 @@
+  use ExUnit.Case
+  doctest Test
+
+  test "execute tests on hardware" do
```
Delete?

fhunleth commented:
A side question on this is that if random people want to provision their Raspberry Pi Zero W's to be part of the regression test farm are you ok with publishing the public key (I guess making it easy to find since it will be in the .fw file for the resourceful person) and letting them poll the test server?

mobileoverlord commented:
As of today that is the case. I plan on adding support to the nerves test server to only allow devices that have been accepted by the server.
